kernel/src/actions/mod.rs

```diff
@@ -481,4 +555,66 @@ mod tests {
         )]));
         assert_eq!(schema, expected);
     }
+
+    #[test]
+    fn test_validate_protocol() {
```

Comment:
Would like to see the feature subset check get exercised. Would also like to see the `min_reader_version == 1` or 2 get checked as well.